src/common/mempool.cc

```diff
@@ -12,21 +12,156 @@
-// because shard[] is defined in the class
-static thread_local size_t thread_shard_index = mempool::num_shards;
-#endif
-
 // default to debug_mode off
 bool mempool::debug_mode = false;
 
 // --------------------------------------------------------------
 
+namespace mempool {
+
+static size_t num_shard_bits;
+static size_t num_shards;
+
+static thread_local size_t thread_shard_index = max_shards;
+
+#if !defined(MEMPOOL_SCHED_GETCPU)
+// There is no cheap way of determining which core a thread is assigned to.
+// Statically assign threads to shards to minimize how many threads use the
+// same shard
+static int thread_shard_next;
+static std::mutex thread_shard_mtx;
+
+#else
+// Use sched_getcpu() to determine which core a thread is running on,
+// use this to assign the thread a shard and cache this value to
+// eliminate overhead of repeatedly calling sched_getcpu. Periodically
+// check the cached value to deal with non-bound threads switching cores.
+// Ideally this means that each CPU core accesses its own shard and there
+// are no shared CPU cachelines. If a thread switches cores performance
+// will be non-optimal (but atomics ensure correctness) until the switch
+// is noticed and a new shard is selected
+
+enum {
+  thread_shard_recheck_min_interval = 1,
+  thread_shard_recheck_max_interval = 10000
+};
+
+static thread_local int thread_shard_recheck_count;
+static thread_local int thread_shard_recheck_interval;
+
+// Keep statistics on how successfully we manage to track which core a thread
+// is running on
+//
+// correct_core - number of times thread to core binding is correctly tracked
+// incorrect_core - number of times thread to core binding is incorrectly tracked
+// checked_core - number of times sched_getcpu() has been called
+//
+// sched_getcpu_cost = 32 cycles
+// cacheline_ping_pong_cost = 150 cycles (intra socket), 400 cycles (inter socket)
+//
+// These costs will vary between CPU architectures, but relative difference between
+// costs is likely to be similar
+//
+// cached_algorithm_cost = checked_core * sched_getcpu_cost + incorrect_core * cacheline_ping_pong_cost
+// uncached_algorithm_cost = (correct_core + incorrect_core) * sched_getcpu_cost
+
+struct fault_stats_t {
+  ceph::atomic<size_t> correct_core = {0};
+  ceph::atomic<size_t> incorrect_core = {0};
+  ceph::atomic<size_t> checked_core = {0};
+
+  char __padding[128 - sizeof(ceph::atomic<size_t>)*2];
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_int("correct_core", correct_core );
+    f->dump_int("incorrect_core", incorrect_core);
+    f->dump_int("checked_core", checked_core);
+  }
+
+  fault_stats_t& operator+=(const fault_stats_t& o) {
+    correct_core += o.correct_core;
+    incorrect_core += o.incorrect_core;
+    checked_core += o.checked_core;
+    return *this;
+  }
+} __attribute__ ((aligned (128)));
+
+static fault_stats_t *shard_faults;
+#endif
+
+}
+
+size_t mempool::get_num_shards(void) {
```

Comment:
Can this function allowed to be called from more than one thread concurrently? I guess no

Reply:
Yes it can.

The very first time its called when num_shards is 0 (hasn't been initialized yet) the code is not multi-thread safe. However this first call will be for one of the static initialization's of a Pool structure that has a dynamic array of size get_num_shards() which is when the code is still single threaded.

I originally tried static initialization to set up num_shards, but this just got into a mess with static initialization ordering problems, so I resorted to initialize on first use.

I''ll add a mutex like this to make the first call thread safe too. As this is performance critical code we only want to take the mutex when initialization is required so it will look like this:

if (num_shards==0) {
    std::lock_guard<std::mutex> lck (thread_shard_mtx);
    if (num_shards==0) {
      // One time initialization
      ...
    }
 }